File: src/main/java/commands/utils/Help.java
package commands.utils;

import com.jagrosh.jdautilities.commons.waiter.EventWaiter;
import com.jagrosh.jdautilities.doc.standard.CommandInfo;
import commandHandler.Command;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import org.jetbrains.annotations.NotNull;
import org.reflections8.Reflections;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.*;

@CommandInfo(name = {"help", "h"}, description = "help command", usage = "help commandname")
public class Help extends Command {
	public ArrayList<Command> cmds = new ArrayList<>();
	EmbedBuilder mainEmbed = new EmbedBuilder();
	Reflections reflections = new Reflections("commands");
	Set<Class<? extends Command>> classes = reflections.getSubTypesOf(Command.class);

	public Help(EventWaiter waiter) throws InstantiationException, IllegalAccessException, NoSuchMethodException, InvocationTargetException {
		super(waiter);
		for (Class<? extends Command> cmd : classes) {
			if ((cmd.getAnnotation(CommandInfo.class).name())[0].equals("Help")) continue;
			Constructor<? extends Command> commandConstructor = cmd.getConstructor(EventWaiter.class);
			Command classInstance = commandConstructor.newInstance(waiter);
			cmds.add(classInstance);
		}
		for (Command cmd : cmds) {
			CommandInfo commandInfo = cmd.getClass().getAnnotation(CommandInfo.class);
			ArrayList<String> aliases = new ArrayList<>();
			Collections.addAll(aliases, commandInfo.name());
			String cmdName = aliases.remove(0);
			mainEmbed
					.setTitle("help")
					.addField(cmdName, commandInfo.description() + (aliases.isEmpty() ? "" : (" - Aliases: `" + String.join("`, `", aliases)) + "`"), true);
		}
	}

	@Override
	protected void execute(@NotNull MessageReceivedEvent event, ArrayList<String> args) {
		if (!args.isEmpty()) {
			EmbedBuilder customHelp = new EmbedBuilder();
			CommandInfo commandInfo = null;
			for (Command cmd : cmds) {
				if (Arrays.asList(cmd.getClass().getAnnotation(CommandInfo.class).name()).contains(args.get(0)))
					commandInfo = cmd.getClass().getAnnotation(CommandInfo.class);
			}
			if (commandInfo != null) {
				List<String> aliases = Arrays.asList(commandInfo.name());
				customHelp
						.setTitle(aliases.get(0))
						.setDescription(commandInfo.description())
						.addField("usage", "`" + commandInfo.usage() + "`", false)
						.addField("aliases", "`" + String.join("`, `", aliases) + "`", false);
				event.getChannel().sendMessage(customHelp.build()).queue();
			} else {
				event.getChannel().sendMessage("enter a valid command!").queue();
			}
		} else {
			event.getChannel().sendMessage(mainEmbed.build()).queue();
		}
	}
}

File: src/main/kotlin/commands/utils/Help.kt
package commands.utils

import com.jagrosh.jdautilities.commons.waiter.EventWaiter
import com.jagrosh.jdautilities.doc.standard.CommandInfo
import com.mongodb.client.MongoClient
import commandHandler.Command
import dev.minn.jda.ktx.await
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import org.reflections8.Reflections
import java.lang.reflect.Constructor
import java.util.*

@CommandInfo(name = ["help", "h"], description = "help command", usage = "help commandname")
class Help constructor(waiter: EventWaiter?, mongoClient: MongoClient) : Command(waiter, mongoClient) {
    private var cmds: ArrayList<Command> = ArrayList()
    private var mainEmbed: EmbedBuilder = EmbedBuilder()
    private var reflections: Reflections = Reflections("commands")
    var classes: Set<Class<out Command>> = reflections.getSubTypesOf(Command::class.java)
    override suspend fun execute(event: MessageReceivedEvent, args: ArrayList<String>) {
        if (args.isNotEmpty()) {
            val customHelp = EmbedBuilder()
            var commandInfo: CommandInfo? = null
            for (cmd: Command in cmds) {
                if (listOf(*cmd.javaClass.getAnnotation(CommandInfo::class.java).name).contains(args[0])) commandInfo = cmd.javaClass.getAnnotation(CommandInfo::class.java)
            }
            if (commandInfo != null) {
                val aliases: List<String> = listOf(*commandInfo.name)
                customHelp
                        .setTitle(aliases[0])
                        .setDescription(commandInfo.description)
                        .addField("usage", "`" + commandInfo.usage + "`", false)
                        .addField("aliases", "`" + java.lang.String.join("`, `", aliases) + "`", false)
                event.channel.sendMessage(customHelp.build()).await()
            } else {
                event.channel.sendMessage("enter a valid command!").await()
            }
        } else {
            event.channel.sendMessage(mainEmbed.build()).await()
        }
    }

    init {
        for (cmd: Class<out Command> in classes) {
            if (((cmd.getAnnotation(CommandInfo::class.java).name)[0] == "help")) continue
            val commandConstructor: Constructor<out Command> = cmd.getConstructor(EventWaiter::class.java, MongoClient::class.java)
            val classInstance: Command = commandConstructor.newInstance(waiter, mongoClient)
            cmds.add(classInstance)
        }
        for (cmd: Command in cmds) {
            val commandInfo: CommandInfo = cmd.javaClass.getAnnotation(CommandInfo::class.java)
            val aliases: ArrayList<String> = ArrayList()
            Collections.addAll(aliases, *commandInfo.name)
            val cmdName: String = aliases.removeAt(0)
            mainEmbed
                    .setTitle("help")
                    .addField(cmdName, commandInfo.description + (if (aliases.isEmpty()) "" else (" - Aliases: `" + java.lang.String.join("`, `", aliases)) + "`"), true)
        }
    }
}